packages/evm/contracts/core/modules/ACL.sol

```diff
@@ -0,0 +1,96 @@
+pragma solidity 0.5.17;
+
+
+contract ACL {
+    string private constant ERROR_BAD_FREEZE = "ACL_BAD_FREEZE";
+    string private constant ERROR_ROLE_GRANTED = "ACL_ROLE_GRANTED";
+    string private constant ERROR_ROLE_NOT_GRANTED = "ACL_ROLE_NOT_GRANTED";
+    string private constant ERROR_ROLE_ALREADY_FROZEN = "ACL_ROLE_ALREADY_FROZEN";
+    string private constant ERROR_INVALID_BULK_INPUT = "ACL_INVALID_BULK_INPUT";
+
+    enum BulkOp { Grant, Revoke, Freeze }
+
+    address internal constant FREEZE_FLAG = address(1);
+    address internal constant ANY_ADDR = address(-1);
+
+    // List of all roles assigned to different addresses
+    mapping (bytes32 => mapping (address => bool)) public roles;
+
+    event Granted(bytes32 indexed id, address indexed who);
+    event Revoked(bytes32 indexed id, address indexed who);
+    event Frozen(bytes32 indexed id);
+
+    modifier notFrozen(bytes32 _id) {
+        require(!isRoleFrozen(_id), ERROR_ROLE_ALREADY_FROZEN);
+        _;
+    }
+
+    /**
+    * @dev Tell whether an address has a role assigned
+    * @param _who Address being queried
+    * @param _id ID of the role being checked
+    * @return True if the requested address has assigned the given role, false otherwise
+    */
+    function hasRole(address _who, bytes32 _id) public view returns (bool) {
+        return roles[_id][_who] || roles[_id][ANY_ADDR];
+    }
+
+    /**
+    * @dev Tell whether a role is frozen
+    * @param _id ID of the role being checked
+    * @return True if the given role is frozen, false otherwise
+    */
+    function isRoleFrozen(bytes32 _id) public view returns (bool) {
+        return roles[_id][FREEZE_FLAG];
+    }
+
+    /**
+    * @dev Internal function to grant a role to a given address
+    * @param _id ID of the role to be granted
+    * @param _who Address to grant the role to
+    */
+    function _grant(bytes32 _id, address _who) internal notFrozen(_id) {
+        require(_who != FREEZE_FLAG, ERROR_BAD_FREEZE);
+        require(!hasRole(_who, _id), ERROR_ROLE_GRANTED);
```

Comment:
I was thinking about this for the Govern ACL too, what do you think about not using a require, but skipping the state change and event emission if the user was already granted the role?